src/hotspot/share/gc/g1/g1ConcurrentRefine.cpp

```diff
@@ -345,86 +200,192 @@ uint G1ConcurrentRefine::max_num_threads() {
   return G1ConcRefinementThreads;
 }
 
-static size_t calc_new_green_zone(size_t green,
-                                  double logged_cards_scan_time,
-                                  size_t processed_logged_cards,
-                                  double goal_ms) {
-  // Adjust green zone based on whether we're meeting the time goal.
-  // Limit to max_green_zone.
-  const double inc_k = 1.1, dec_k = 0.9;
-  if (logged_cards_scan_time > goal_ms) {
-    if (green > 0) {
-      green = static_cast<size_t>(green * dec_k);
+void G1ConcurrentRefine::update_pending_cards_target(double logged_cards_scan_time_ms,
```

Comment:
`logged_cards_scan_time_ms` includes time for merging card (i think it should include time for concatenation),  probably better to name it as `logged_cards_processing_time_ms` as in the caller. Additionally, we use `processed_logged_cards` and not `scanned_logged_cards`

Reply:
Agreed that "scan" isn't right.  I dropped that word in several places.  Just
dropped rather than replacing with "processing" because of excessive length
and I didn't think it added anything.